jupyter_console/interactiveshell.py

```diff
@@ -597,3 +637,21 @@ def init_history(self):
         self.history_manager = ZMQHistoryManager(client=self.client)
         self.configurables.append(self.history_manager)
 
+    def get_source_and_reset(self):
+        """
+        get and reset source input line(s)
+        """
+        source, self._source = self._source, ''
+        return source
+
+    def handle_source(self, line):
+        """
+        more, indent
+        """
+        ## Ask client if line is complete; get indent for next line:
+        if self._source:
+            self._source += "\n"
```

Comment:
I think it would be neater to store the input as a list, and then call `self.client.is_complete('\n'.join(self._source_lines_buffered))`